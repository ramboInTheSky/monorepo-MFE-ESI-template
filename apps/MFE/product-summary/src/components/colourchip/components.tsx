import styled from "styled-components"
import { Link } from "@mui/material"

export const ColourChipListItem = styled.li`
    box-sizing: border-box;
    margin-right: 0.125rem;
    margin-bottom: 0.125rem;
    border: 0.125rem solid #fff;
    border-radius: ${({ theme }) => theme.colours.utilities.chipBorderRadius};
    max-height: 2.25rem;
    max-width: 2.25rem;
    width: calc((100% - 1rem) / 9);
    border-width: 0.0625rem;
    position: relative;

    &.selected {
        border: ${props => props.theme.colours.utilities.activeBorder};

        &:before {
            border-radius: ${({ theme }) => `calc(${theme.colours.utilities.chipBorderRadius} - .125rem)`};
            border: 0.0625rem solid #fff;
            box-sizing: border-box;
            content: "";
            display: block;
            height: 100%;
            position: absolute;
            pointer-events: none;
            width: 100%;
            z-index: 1;
        }
    }
`

export const ColourChipLink = styled(Link)`
    padding-top: 100%;
    border-radius: ${props => props.theme.colours.utilities.chipBorderRadius};
    position: relative;
    display: block;
    // Color value to be part of theme object - PBI23328
    background: #d1d1d1;

    > img {
        max-height: 100%;
        max-width: 100%;
        border-radius: ${props => props.theme.colours.utilities.chipBorderRadius};
        position: absolute;
        top: 0;
        left: 0;
    }
`
