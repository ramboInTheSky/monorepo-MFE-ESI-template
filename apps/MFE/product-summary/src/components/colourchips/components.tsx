import { Collapse } from '@mui/material';
import {breakpoints} from "@monorepo/themes"
import styled from "styled-components"

export const ColourChipCollapse = styled(Collapse)`
    margin-top: 0.5rem;
    margin-bottom: 0.5rem;
`

export const ColourChipViewButtonContainer = styled.div``

export const ColourChipsList = styled.ul`
    list-style-type: none;
    display: flex;
    flex-wrap: wrap;
    padding: 0;
    overflow: hidden;
    max-width: 13.5rem;
    margin: 0;
    ${({theme}) =>
        theme.enablingCenteredContent &&
        `
        justify-content: center;
        margin-left: 0.75rem;
    `}
    > li:nth-child(9n) {
        margin-right: 0;
    }
`

export const ColourChipsContainer = styled.div`
    @media (max-width: ${breakpoints.values.lg - 1}px) {
        display: none;
    }

    > div:nth-child(2n) {
        display: ${({theme}) => `${theme.enablingCenteredContent ? "inline-flex" : "flex"}`};
        margin-top: 0 0.5rem 0 0.5rem;
        cursor: pointer;
    }

    > div:nth-child(2n) > button {
        border: 0;
        padding: 0;
        background: none;
        margin-right: 0.5rem;
        color: ${props => props.theme.colours.text.hyperlink};
        font-family: ${({theme}) => `${theme.colours.font.primary.medium.family}, ${theme.colours.font.default}`};
        margin-right: 0.5rem;
        font-size: 0.875rem;
        cursor: pointer;
        line-height: 1.5;

        > img {
            margin-left: 0.5rem;
        }
    }
`
