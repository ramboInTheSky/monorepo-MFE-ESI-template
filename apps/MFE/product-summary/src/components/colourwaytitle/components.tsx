import styled from "styled-components"
import {Link} from "@mui/material"
import {breakpoints} from "@monorepo/themes"

export const TileTitle = styled(Link)`
    height: 1.5em;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    display: block;
    text-transform: capitalize;
    text-decoration: none;

    @media (max-width: ${breakpoints.values.lg - 1}px) {
        height: 3em;
        white-space: normal;
        -webkit-line-clamp: 2;
        -webkit-box-orient: vertical;
        display: -webkit-box;
        margin-bottom: 0.25rem;

        &:after {
            position: relative;
            content: "";
            border-bottom: 0.0625rem solid #d1d1d1;
            width: 1rem;
            margin: auto;
            display: block;
            bottom: -0.8rem;
        }
    }
`
