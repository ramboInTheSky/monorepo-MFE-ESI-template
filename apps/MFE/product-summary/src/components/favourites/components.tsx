import styled from "styled-components"
import {FavouriteState} from "../../models/Favourites"
import env from "../../config/env"

interface FavouriteProps {
    favState: string
    shouldAnimate: boolean
}

export const FavouritesContainer = styled.div`
    position: absolute;
    bottom: 1.125rem;
    right: 0.625rem;
    border: ${props => props.theme.colours.utilities.dividerDark};
    background: #fff;
    border-radius: 50%;
    cursor: pointer;
    z-index: 100;
    height: 2.25rem;
    width: 2.25rem;
`

export const FavouritesIcon = styled.svg<FavouriteProps>`
    && {
        /* transform:rotate added to prevent pixel snapping and rectify visible jump in animation*/
        /* https://stackoverflow.com/questions/22097660/is-there-a-way-i-can-force-chrome-to-do-subpixel-rendering-for-a-slow-translatio
https://johnresig.com/blog/sub-pixel-problems-in-css/ */
        transform: rotateZ(-0.0000000001deg);
        width: 100%;
        height: 100%;

        ${AnimationProps =>
            AnimationProps.favState === FavouriteState.Inactive &&
            `

                @media all and (-ms-high-contrast: none) {
                    background-image: url(${env.REACT_APP_BLOB_STORAGE_PATH}/icons/shared/favourites-inactive.png); 
                    background-repeat: no-repeat;
                    background-size: 100% 100%;
                }
        `}

        ${FavouriteProps =>
            FavouriteProps.favState === FavouriteState.Loading &&
            `
            stroke-dasharray: 40;
            stroke-dashoffset: 80;
            stroke-linecap: round;
            animation: favourite-loading 0.8s linear infinite reverse;

            @keyframes favourite-loading {
                0% {
                    stroke-dashoffset: 80;
                }
                100% {
                    stroke-dashoffset: 160;
                }
            }

            @media all and (-ms-high-contrast: none) {
                background-image: url(${env.REACT_APP_BLOB_STORAGE_PATH}/icons/shared/favourites-loading.gif); 
                background-repeat: no-repeat;
                background-size: 100% 100%;
            }
        `}

        ${FavouriteProps =>
            FavouriteProps.favState === FavouriteState.Active &&
            FavouriteProps.shouldAnimate &&
            `
            animation: favourite-loaded 1s ease-in-out;
            
            @keyframes favourite-loaded {
                0% {
                    transform: scale(1);
                }
                50% {
                    transform: scale(1.25);
                }
                100% {
                    transform: scale(1);
                }
            }
            
        `}

        ${FavouriteProps =>
            FavouriteProps.favState === FavouriteState.Active &&
            `
            @media all and (-ms-high-contrast: none) {
                background-image: url(${env.REACT_APP_BLOB_STORAGE_PATH}/icons/shared/favourites-active.png); 
                background-repeat: no-repeat;
                background-size: 100% 100%;
            }
            
        `}
    }
`

export const FavouritesButton = styled.button`
    && {
        padding: 0.5rem 0.5rem;
        height: 100%;
        width: 100%;
        background: none;
        border: none;
        cursor: pointer;
    }
`
