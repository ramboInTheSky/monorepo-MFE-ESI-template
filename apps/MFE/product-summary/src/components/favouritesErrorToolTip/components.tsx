import styled from "styled-components"

const buttonOverrides = `
    -webkit-appearance: none;
    appearance: none;
    background: none;
    border: 0;
    cursor: pointer;
    padding: 0;
`

export const Container = styled.div`
    padding: 1rem;
    width: 18rem;
    border: ${props => props.theme.colours.popover.border};
    background: ${props => props.theme.colours.header.myAccount.background};
    border-radius: ${props => props.theme.colours.popover.radius};
    box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.4);
`

export const ErrorContainer = styled(Container)`
    width: 8.9rem;
    text-align: center;
`

export const Header = styled.div`
    margin-bottom: 1rem;
    font-family: ${({theme}) => `${theme.colours.font.primary.regular.family}, ${theme.colours.font.default}`};
    display: flex;
    justify-content: space-between;
`

export const Title = styled.div`
    font-family: ${({theme}) => `${theme.colours.font.primary.medium.family}, ${theme.colours.font.default}`};
    font-size: 0.938rem;
    font-weight: 500;
`

export const CloseButton = styled.button`
    ${buttonOverrides}
    height: 0.75rem;
    width: 0.75rem;
`

export const CloseImage = styled.img`
    width: 0.75rem;
    height: 0.75rem;
`

export const Content = styled.div`
    font-size: 0.875rem;
    line-height: 1.313rem;
`

export const ShoppingLinkWrapper = styled.div`
    text-align: center;
`

export const ShoppingLink = styled.button`
    ${buttonOverrides}
    color: ${props => props.theme.colours.text.hyperlink};
    font-family: ${({theme}) => `${theme.colours.font.primary.regular.family}, ${theme.colours.font.default}`};
    font-size: 0.875rem;
    text-decoration: underline;
`

export const ManageFavouritesButton = styled.div`
    height: 2.25rem;
    margin: 1.1rem 0 1.1rem;
    & > a {
        text-decoration: none;
        width: 100%;
        display: flex;
        justify-content: center;
        align-items: center;
        padding: 0 !important;
        font-size: 1rem !important;
        font-weight: 500 !important;
        visibility: visible !important;
    }
`