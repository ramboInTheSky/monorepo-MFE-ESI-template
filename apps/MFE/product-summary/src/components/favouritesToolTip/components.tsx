import styled from "styled-components"

export const Container = styled.div`
    padding: 1rem;
    min-width: 18rem;
    border: ${props => props.theme.colours.popover.border};
    background: ${props => props.theme.colours.header.myAccount.background};
    border-radius: ${props => props.theme.colours.popover.radius};
    box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.4);
`

export const Header = styled.div`
    margin-bottom: 1rem;
    font-family: ${({theme}) => `${theme.colours.font.primary.regular.family}, ${theme.colours.font.default}`};
    display: flex;
    justify-content: space-between;
`

export const Title = styled.div`
    font-size: 0.938rem;
`

export const CloseButton = styled.button`
    -webkit-appearance: none;
    appearance: none;
    background: none;
    border: 0;
    cursor: pointer;
    padding: 0;
    height: 0.75rem;
    width: 0.75rem;
`

export const CloseImage = styled.img`
    width: 0.75rem;
    height: 0.75rem;
`

export const Content = styled.div`
    font-size: 0.875rem;
`

export const Link = styled.a`
    color: ${props => props.theme.colours.text.hyperlink};
`

export const UserSection = styled.p`
    margin-bottom: 0.2rem;
`
