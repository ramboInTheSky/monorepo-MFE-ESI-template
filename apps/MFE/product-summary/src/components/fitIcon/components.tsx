import styled from "styled-components"

export const StyledImage = styled.img`
    margin-bottom: 0.5rem;
    width: 1rem;
    height: 1rem;
`

export const FitsContainer = styled.div`
    position: absolute;
    left: 0.5rem;
    bottom: 0rem;
    max-width: 1rem;
    z-index: 2;
`
