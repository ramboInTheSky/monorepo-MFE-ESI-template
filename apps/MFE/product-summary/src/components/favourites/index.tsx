import React, {useEffect, useRef, useState} from "react"
import {FavouriteState} from "../../models/Favourites"
import {FavouritesContainer, FavouritesIcon, FavouritesButton} from "./components"
import connect from "./connect"
import {useProductFavourites} from "../../hooks/useProductFavourites"
import {
    getFavouriteIconPath,
    shoppingListItemsToFavouritedColourways,
    setFavoritesCallback,
    getNextFavouritesData,
} from "../../utils/favourites"
import {FavErrorToolTipType} from "../../models/ProductSummary"
import FavouritesErrorToolTip from "../favouritesErrorToolTip"
import FavouritesToolTip from "../favouritesToolTip"
import {getFavourtiesSubscriptionCallBack} from "../../events"
import {TextModel} from "../../models/Text"


interface ProductFavouritesProps {
    addToFavourites: () => void
    removeFromFavourites: () => void
    baseUrl: string
    isFav: boolean
    itemNumber: string
    showFavErrorToolTip: FavErrorToolTipType | null
    isLoadingFav: boolean
    text: TextModel
    setFavouritedColourways: (favoritedColourways: string[]) => void
    animateFavouriteIcon: boolean
    selectedColourwayTitle: string
}

export const ProductFavourites = ({
    isFav,
    addToFavourites,
    removeFromFavourites,
    baseUrl,
    setFavouritedColourways,
    isLoadingFav,
    text,
    animateFavouriteIcon,
    selectedColourwayTitle
}: ProductFavouritesProps) => {
    const [favClicked, setFavClicked] = useState(false)

    const {toggleProductFavourites, favState} = useProductFavourites({
        isFav,
        addToFavourites,
        removeFromFavourites,
        isLoadingFav,
        setFavClicked
    })

    useEffect(() => {
        const favouritesData = getNextFavouritesData()
        if (favouritesData) {
            setFavouritedColourways(shoppingListItemsToFavouritedColourways(favouritesData?.ShoppingListItems))
        } else {
            getFavourtiesSubscriptionCallBack(setFavoritesCallback, setFavouritedColourways)
        }
    }, [])

    const favoritesIconRef = useRef(null)

    const getTranslatedFavState = (favouriteState: FavouriteState) => {
        const {inactive, loading, active} = text.favouriteStates
        switch (favouriteState) {
            case FavouriteState.Active:
                return active
            case FavouriteState.Inactive:
                return inactive
            case FavouriteState.Loading:
                return loading
            default:
                return ""
        }
    }

    const formatButtonTitleAndLabel = `${text.product.favourites} ${selectedColourwayTitle} ${getTranslatedFavState(favState)}`

    return (
        <FavouritesContainer
            className="product-summary-favourites-container"
            data-testid="product-summary-favourites-button-container"
        >
            <FavouritesButton
                onClick={toggleProductFavourites}
                data-testid="product-summary-favourites-button"
                aria-label={formatButtonTitleAndLabel}
                title={formatButtonTitleAndLabel}
            >
                <FavouritesIcon
                    id="product-summary-favourites-icon"
                    data-testid={`product-${favState}-favourite-icon`}
                    favState={favState}
                    ref={favoritesIconRef}
                    shouldAnimate={animateFavouriteIcon}
                >
                    <title>{formatButtonTitleAndLabel}</title>
                    <use
                        xmlnsXlink="http://www.w3.org/1999/xlink"
                        xlinkHref={getFavouriteIconPath(favState)}
                        viewBox="0 0 26 24"
                    />
                </FavouritesIcon>
            </FavouritesButton>            
            <FavouritesToolTip
                isProductFavourited={isFav && favClicked}
                favState={favState}
                baseUrl={baseUrl}
                referenceElement={favoritesIconRef.current}
                text={text}
                setFavClicked={setFavClicked}
            />
            <FavouritesErrorToolTip referenceElement={favoritesIconRef.current} />
        </FavouritesContainer>
    )
}

export default connect(ProductFavourites)
