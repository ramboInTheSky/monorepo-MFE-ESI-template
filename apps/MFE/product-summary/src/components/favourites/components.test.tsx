import React from "react"
import {render} from "@testing-library/react"
import {ThemeProvider} from "styled-components"
import {mockTheme} from "../../../__mocks__/mockStore"
import {FavouritesContainer, FavouritesIcon} from "./components"

describe("Given a FavouritesContainer - FavouritesContainer", () => {
    it("should render as expected when it is Active", () => {
        const {asFragment} = render(
            <ThemeProvider theme={mockTheme}>
                <FavouritesContainer />
            </ThemeProvider>,
        )
        expect(asFragment()).toMatchSnapshot()
    })
    it("should render as expected  when it is Loading", () => {
        const {asFragment} = render(
            <ThemeProvider theme={mockTheme}>
                <FavouritesContainer />
            </ThemeProvider>,
        )
        expect(asFragment()).toMatchSnapshot()
    })
})

describe("Given a FavouritesIcon - FavouritesIcon", () => {
    it("should render as expected when favState is inactive", () => {
        const {asFragment} = render(<FavouritesIcon favState="inactive" shouldAnimate/>)
        expect(asFragment()).toMatchSnapshot()
    })
    it("should render as expected when favState is loading", () => {
        const {asFragment} = render(<FavouritesIcon favState="loading" shouldAnimate/>)
        expect(asFragment()).toMatchSnapshot()
    })
    it("should render as expected when favState is active", () => {
        const {asFragment} = render(<FavouritesIcon favState="active" shouldAnimate/>)
        expect(asFragment()).toMatchSnapshot()
    })
})
