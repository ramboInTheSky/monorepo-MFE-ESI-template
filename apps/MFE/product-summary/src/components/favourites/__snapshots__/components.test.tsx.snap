// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a FavouritesContainer - FavouritesContainer should render as expected  when it is Loading 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
  bottom: 1.125rem;
  right: 0.625rem;
  border: .0625rem solid #9e9e9e;
  background: #fff;
  border-radius: 50%;
  cursor: pointer;
  z-index: 100;
  height: 2.25rem;
  width: 2.25rem;
}

<div
    class="c0"
  />
</DocumentFragment>
`;

exports[`Given a FavouritesContainer - FavouritesContainer should render as expected when it is Active 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
  bottom: 1.125rem;
  right: 0.625rem;
  border: .0625rem solid #9e9e9e;
  background: #fff;
  border-radius: 50%;
  cursor: pointer;
  z-index: 100;
  height: 2.25rem;
  width: 2.25rem;
}

<div
    class="c0"
  />
</DocumentFragment>
`;

exports[`Given a FavouritesIcon - FavouritesIcon should render as expected when favState is active 1`] = `
<DocumentFragment>
  .c0.c0 {
  -webkit-transform: rotateZ(-0.0000000001deg);
  -ms-transform: rotateZ(-0.0000000001deg);
  transform: rotateZ(-0.0000000001deg);
  width: 100%;
  height: 100%;
  -webkit-animation: favourite-loaded 1s ease-in-out;
  animation: favourite-loaded 1s ease-in-out;
}

@media all and (-ms-high-contrast:none) {
  .c0.c0 {
    background-image: url(/static-content/icons/shared/favourites-active.png);
    background-repeat: no-repeat;
    background-size: 100% 100%;
  }
}

<svg
    class="c0"
  />
</DocumentFragment>
`;

exports[`Given a FavouritesIcon - FavouritesIcon should render as expected when favState is inactive 1`] = `
<DocumentFragment>
  .c0.c0 {
  -webkit-transform: rotateZ(-0.0000000001deg);
  -ms-transform: rotateZ(-0.0000000001deg);
  transform: rotateZ(-0.0000000001deg);
  width: 100%;
  height: 100%;
}

@media all and (-ms-high-contrast:none) {
  .c0.c0 {
    background-image: url(/static-content/icons/shared/favourites-inactive.png);
    background-repeat: no-repeat;
    background-size: 100% 100%;
  }
}

<svg
    class="c0"
  />
</DocumentFragment>
`;

exports[`Given a FavouritesIcon - FavouritesIcon should render as expected when favState is loading 1`] = `
<DocumentFragment>
  .c0.c0 {
  -webkit-transform: rotateZ(-0.0000000001deg);
  -ms-transform: rotateZ(-0.0000000001deg);
  transform: rotateZ(-0.0000000001deg);
  width: 100%;
  height: 100%;
  stroke-dasharray: 40;
  stroke-dashoffset: 80;
  stroke-linecap: round;
  -webkit-animation: favourite-loading 0.8s linear infinite reverse;
  animation: favourite-loading 0.8s linear infinite reverse;
}

@media all and (-ms-high-contrast:none) {
  .c0.c0 {
    background-image: url(/static-content/icons/shared/favourites-loading.gif);
    background-repeat: no-repeat;
    background-size: 100% 100%;
  }
}

<svg
    class="c0"
  />
</DocumentFragment>
`;
